### PR TITLE
Improve 'sysand info' command output

### LINKED ISSUE #208: Better `sysand info` output for multiple projects/versions
I didn't observe douplicated output from `sysand info` on a local project just initialized.

```
sysand info --iri urn:kpar:sysmod
warning: Direct or transitive usages of SysML v2/KerML standard library packages are
         ignored by default. If you want to process them, pass `--include-std` flag
Name: SYSMOD
Description: The library provides concepts of the common and general purpose MBSE methodology SYSMOD.
Version: 5.0.0-alpha.2
License: Apache-2.0
Website: https://mbse4u.com/sysmod
Maintainer(s): MBSE4U, Tim Weilkiens <tim@mbse4u.com>
Topics: MBSE, Methodology
Usages:
Name: SYSMOD
Description: The library provides concepts of the common and general purpose MBSE methodology SYSMOD.
Version: 5.0.0-alpha.1
License: Apache-2.0
Website: https://mbse4u.com/sysmod
Maintainer(s): MBSE4U, Tim Weilkiens <tim@mbse4u.com>
Topics: MBSE, Methodology
Usages:
```

### PR DESCRIPTION
**Pull request by Jonas-Puksta-Sensmetry**
Resolves #208.

Changes:
* Make sysand info return only the info of the highest package version
* Make sysand path info command in the bindings consider missing info or meta an error
* Improve the sysand info printing - colorize output and be more explicit about skipped standard library usages

### CODE REVIEW

**Comment by consideRatio** [comment]
Oh nice picking this up @Jonas-Puksta-Sensmetry!!

I'm on mobile and asked opus 4.7 to review with this prompt:

> https://github.com/sensmetry/sysand/pull/292#event-24609423150
>
>Task three Staff SWE to review the linked PR separately. Let one focus on avoiding bugs etc, let another focus on architecture decisions, and another one on code comprehension (comments, naming, etc). Then summarize for me.

(I havn't considered this PR or the AI review myself yet)
---

Got the diff — single-file change to `sysand/src/commands/info.rs` (+51/-21). Here's what each reviewer flagged:

**Reviewer 1 — Bugs & correctness**
- Behavioral regression: if *no* project in the URI response has a valid semver version, the command prints nothing to stdout (only a `log::warn!`). Old code would have printed all of them. A user running `sysand info <uri>` may see silent success.
- Warning spam: the loop logs `"Version X is not a valid semantic version"` once per invalid entry — potentially noisy on large resolvers.
- Tie-breaking: uses `>` not `>=`, so the first project at the max version wins. If `do_info` ordering isn't deterministic, output becomes non-deterministic on ties.
- The three new `return;` statements inside `pprint_interchange_project` are safe today (usages block is last), but brittle — any future field printed after usages will be silently skipped for the "empty / all-ignored" cases.

**Reviewer 2 — Architecture**
- Selection logic leaks into the CLI layer. "Pick highest semver" belongs in `sysand_core` (or a helper), not inlined in `command_info_uri`. Makes it untestable and creates divergence with `command_info_path`, which has no such selection.
- Asymmetry between `command_info_path` (prints one) and `command_info_uri` (now filters to one) deserves a comment or consolidation — why does only the URI variant need picking?
- The "Some usages are ignored" / "All usages are ignored" strings mix UX with data. Fine for now, but a structured output mode (JSON) will need to rethink this.
- No tests added for version-selection edge cases (empty, all-invalid, ties, mixed valid/invalid). This logic is subtle enough to warrant a few unit tests.
- Good call on `&InterchangeProjectInfoRaw` — removes needless moves and enables the selection loop to hold references cheaply.

**Reviewer 3 — Comprehension (naming, comments, clarity)**
- `best_version_info` is vague ("best" how?). `highest_version_project` or `latest_project` reads better.
- `is_usage_ignored` is singular but means "any usage is ignored" → rename `any_usages_ignored` / `has_ignored_usages`.
- String inconsistency: `"No usages."` has a period, the two new messages don't.
- `log::warn!("No semantic versions found")` is ambiguous — sounds like nothing resolved. Prefer: "No projects with valid semantic versions".
- Invalid-version warning doesn't identify *which* project — add the name/IRI.
- `usages_to_print.iter()` on a `Vec<&Usage>` gives `&&Usage`; just iterate the vec directly.
- No doc comment explaining *why* "highest semver" is the selection policy — worth one line.

**Summary / what I'd push back on before merging**
The change does what the PR says, and the refactor to borrow `&InterchangeProjectInfoRaw` is clean. Three things are worth addressing before approval: (1) the silent-stdout case when every version fails semver parsing is a real behavioral regression — either fall back to printing something or surface a proper error; (2) move version selection into core with unit tests; (3) the naming/message polish from reviewer 3 is cheap and noticeably improves the command's UX. The rest (ties, warning volume, early-return brittleness) are nits worth a comment but not blockers.

**Comment by consideRatio** [comment]
Screenshot_20260417-164944

Also the same AI...

**Comment by victor-linroth-sensmetry** [comment]
If you're tidying up the info printer, you could also consider colorizing some of the output.

C.f. Cargo
Screenshot 2026-04-22 at 13 25 30

**Comment by Jonas-Puksta-Sensmetry** [comment]
> If you're tidying up the info printer, you could also consider colorizing some of the output.

Done

**Comment by Jonas-Puksta-Sensmetry** [comment]
I addressed the above comments

**Review by consideRatio — COMMENTED** [review]
I haven't onboarded fully to the code base, but my expectation is that:
- we have the core rust library as the foundation
- we have four surfaces (CLI, bindings x3) exposing foundation parts
- we typically just expose the foundation parts required for CLI commands
- we don't add anything in the surfaces implementations besides the minimal thing to adjust to the specific surface

(I've opened #313 related to aligning on expectations, its beyond the scope of this specific PR.)

Based on that, my expectation was that the change of picking the latest semver version should be a core library change primarily, with surfaces align around it. However, this PR only changes the CLI and doesn't touch the sysand_core module.

Currently this PR touches just the CLI, and specifically for when doing `sysand info --iri` (via `command_info_uri`) rather than for example `sysand info --iri ... version` (via `command_info_verb_uri`).

```
$ sysand info --iri urn:kpar:sysmod
warning: Direct or transitive usages of SysML v2/KerML standard library packages are
         ignored by default. If you want to process them, pass `--include-std` flag
Name: SYSMOD
Description: The library provides concepts of the common and general purpose MBSE methodology SYSMOD.
Version: 5.0.0-alpha.2
License: Apache-2.0
Website: https://mbse4u.com/sysmod
Maintainer(s): MBSE4U, Tim Weilkiens <tim@mbse4u.com>
Topics: MBSE, Methodology
All usages are ignored

$ sysand info --iri urn:kpar:sysmod version
5.0.0-alpha.2
5.0.0-alpha.1
```

Are we aligned on the intent of what we want to accomplish related to this PR? And as a followup to alignment, do we still want to invest time and effort towards it now?

**Comment by andrius-puksta-sensmetry** [comment]
> Based on that, my expectation was that the change of picking the latest semver version should be a core library change primarily, with surfaces align around it.

This also matches my expectation. Specifically, changes should go into `do_info`, and it should be changed to return a single pair of info and meta.

EDIT: I think this does not match what we discussed previously with Jonas, but we should avoid having special behaviors just in CLI.

File: core/src/utils.rs [hunk 1]
```diff
@@ -0,0 +1,13 @@
+// SPDX-License-Identifier: MIT OR Apache-2.0
+// SPDX-FileCopyrightText: © 2026 Sysand contributors <opensource@sensmetry.com>
+
+use std::error::Error;
+
+pub fn format_sources(mut error: &dyn Error) -> String {
+    let mut message = error.to_string();
+    while let Some(source) = error.source() {
+        message += &format!("  caused by: {source}\n");
```
Comment (andrius-puksta-sensmetry) [inline core/src/utils.rs]:
```suggestion
        writeln!(&mut message, "  caused by: {source}").unwrap();
```

File: core/src/commands/info.rs [hunk 2]
```diff
@@ -80,11 +75,41 @@ pub fn do_info<S: AsRef<str>, R: ResolveRead>(
...
+                        log::info!("{}", format_sources(&err));
+                    }
                 };
             }
-            Ok(result)
+            match best_version_info_meta {
+                Some((_, info, meta)) => {
+                    if !non_semantic_versions.is_empty() {
+                        log::warn!(
+                            "The following versions were skipped as they are not semantic versions {}",
```
Comment (andrius-puksta-sensmetry) [inline core/src/commands/info.rs]:
```suggestion
                            "the following versions were skipped as they are not semantic versions {}",
```

File: core/src/commands/info.rs [hunk 3]
```diff
@@ -80,11 +75,41 @@ pub fn do_info<S: AsRef<str>, R: ResolveRead>(
...
+                                (Ok(_), Some(_)) => best_version_info_meta,
+                                (Ok(cur_version), None) => Some((cur_version, info, meta)),
+                                (Err(_), _) => {
+                                    non_semantic_versions.push(info.version);
+                                    best_version_info_meta
+                                }
+                            }
+                    }
+                    Err(err) => {
+                        log::warn!("Ignoring a project because: {err}");
```
Comment (andrius-puksta-sensmetry) [inline core/src/commands/info.rs]:
```suggestion
                        log::warn!("ignoring a project because: {err}");
```

**Comment by andrius-puksta-sensmetry** [comment]
Thinking more about intended use cases:

- CLI: quick lookup of a project. Output should be brief.
- bindings: building integrations into other applications, for example, an IDE that allows searching for packages. I think there it makes sense to have multiple versions, and especially important would be to list all available package sources. We should leave the version filtering up to the caller, as it's not for us to decide whether they want e.g. pre-release versions. But sorting them on our end is useful, as it's most likely that latest package versions would be desired, so this allows callers to not worry about semver sorting.

I think this PR is fine, until we get actual users of the bindings.

File: sysand/src/commands/info.rs [hunk 4]
```diff
@@ -34,43 +35,60 @@ use sysand_core::{
...
-    info: InterchangeProjectInfoRaw,
+    info: &InterchangeProjectInfoRaw,
     excluded_iris: &HashSet<String>,
 ) {
-    println!("Name: {}", info.name);
-    if let Some(publisher) = info.publisher {
-        println!("Publisher: {}", publisher);
+    let header = style::get_style_config().header;
+    println!("{header}Name:{header:#} {}", info.name);
+    if let Some(ref publisher) = info.publisher {
```
Comment (andrius-puksta-sensmetry) [inline sysand/src/commands/info.rs]:
```suggestion
    if let Some(publisher) = &info.publisher {
```
The same for other `ref` usages

File: sysand/src/commands/info.rs [hunk 5]
```diff
@@ -34,43 +35,60 @@ use sysand_core::{
...
+                    println!();
+                }
             }
-            print!("    {}", usage.resource);
-            if let Some(v) = usage.version_constraint {
-                println!(" ({})", v);
-            } else {
-                println!();
+            if has_ignored_usages {
+                println!("Some usages are ignored");
```
Comment (andrius-puksta-sensmetry) [inline sysand/src/commands/info.rs:91]:
```suggestion        let mut has_ignored_usages = false;
        let mut to_print = String::new();
        for usage in &info.usage {
            if !excluded_iris.contains(&usage.resource) {
                write!(&mut to_print, "    {}", usage.resource).unwrap();
                if let Some(ref v) = usage.version_constraint {
                    writeln!(&mut to_print, " ({})", v).unwrap();
                } else {
                    writeln!(&mut to_print).unwrap();
                }
            } else {
                has_ignored_usages = true;
            }
        }
        if !to_print.is_empty() {
            println!("{header}Usages:{header:#}\n{to_print}");
            if has_ignored_usages {
                println!("Some usages are ignored");
            }
        } else {
            if has_ignored_usages {
                println!("All usages are ignored");
```

Comment (Jonas-Puksta-Sensmetry) [reply sysand/src/commands/info.rs:91]: My original code isn't the most elegant, but this is strange too. I'll sick with the original

**Review by andrius-puksta-sensmetry — APPROVED** [review]

File: core/src/utils.rs [hunk 6]
```diff
@@ -0,0 +1,13 @@
+// SPDX-License-Identifier: MIT OR Apache-2.0
+// SPDX-FileCopyrightText: © 2026 Sysand contributors <opensource@sensmetry.com>
+
+use std::{error::Error, fmt::Write as _};
+
+pub fn format_sources(mut error: &dyn Error) -> String {
```
Comment (consideRatio) [inline core/src/utils.rs]: This function is currently used once by core/src/commands/info.rs, in the same crate.

I think its premature to both separate this out as a general utility function and to make it public. If there is a need in the future to use this outside the crate, we should make it public then, and potentially extract it to a dedicated module.

**Review by consideRatio — APPROVED** [review]
Wieee nice @Jonas-Puksta-Sensmetry!!!

I think its worth addressing my one review comment, but I don't consider it blocking.

I opened https://github.com/sensmetry/sysand/issues/327 with something I think we should improve later at some point.

File: core/src/utils.rs (hunk 6 above)
Comment (Jonas-Puksta-Sensmetry) [reply core/src/utils.rs]: Good point - I changed it to `pub(crate)` instead

Comment (andrius-puksta-sensmetry) [reply core/src/utils.rs]: I think it's not worth worrying about what functionality we expose, as we control all the consumers of `sysand_core`.
